File: macos-client/Sources/U/Backend/BackendClient.swift
import Foundation

/// El único puente con el cerebro: `POST {backendUrl}/api/agent/turn`. El cliente manda el estado de
/// pantalla y recibe las acciones a ejecutar. Sin este endpoint, el cliente no sabe pensar — a propósito.
final class BackendClient {
    private let baseUrl: String
    private let userId: String
    private let token: String?

    init(config: Config) {
        baseUrl = config.backendUrl.hasSuffix("/") ? String(config.backendUrl.dropLast()) : config.backendUrl
        userId = config.userId
        token = (config.clientToken?.isEmpty == false) ? config.clientToken : nil
    }

    func turn(_ req: TurnRequest) async throws -> TurnResponse {
        var r = req
        r.userId = userId
        var request = URLRequest(url: URL(string: "\(baseUrl)/api/agent/turn")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        request.httpBody = try JSONEncoder().encode(r)
        request.timeoutInterval = 300

        let (data, response) = try await URLSession.shared.data(for: request)
        let parsed = try JSONDecoder().decode(TurnResponse.self, from: data)
        let code = (response as? HTTPURLResponse)?.statusCode ?? 0
        if code >= 300 {
            throw NSError(domain: "U", code: code, userInfo: [NSLocalizedDescriptionKey: parsed.error ?? "backend HTTP \(code)"])
        }
        return parsed
    }
}

File: windows-client/src/Backend/BackendClient.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using U.WindowsClient.Domain;

namespace U.WindowsClient.Backend;

/// <summary>
/// El único puente con el cerebro: <c>POST {BackendUrl}/api/agent/turn</c>. El cliente manda el estado
/// de pantalla y recibe las acciones a ejecutar. No hay otra llamada de red con inteligencia: si este
/// endpoint no responde, el cliente no sabe pensar por su cuenta — a propósito.
/// </summary>
public sealed class BackendClient
{
    private readonly HttpClient _http;
    private readonly string _baseUrl;
    private readonly string _userId;

    private static readonly JsonSerializerOptions Json = new()
    {
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
    };

    public BackendClient(Config config)
    {
        _baseUrl = config.BackendUrl.TrimEnd('/');
        _userId = config.UserId;
        _http = new HttpClient { Timeout = TimeSpan.FromMinutes(5) };
        if (!string.IsNullOrWhiteSpace(config.ClientToken))
            _http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", config.ClientToken);
    }

    public async Task<TurnResponse> TurnAsync(TurnRequest req, CancellationToken ct)
    {
        req.UserId = _userId;
        var body = JsonSerializer.Serialize(req, Json);
        using var content = new StringContent(body, Encoding.UTF8, "application/json");
        using var res = await _http.PostAsync($"{_baseUrl}/api/agent/turn", content, ct);
        var text = await res.Content.ReadAsStringAsync(ct);
        var parsed = JsonSerializer.Deserialize<TurnResponse>(text, Json);
        if (parsed == null) throw new InvalidOperationException($"respuesta vacía del backend (HTTP {(int)res.StatusCode})");
        if (!res.IsSuccessStatusCode)
            throw new InvalidOperationException(parsed.Error ?? $"backend HTTP {(int)res.StatusCode}");
        return parsed;
    }

    /// <summary>POST genérico hacia cualquier endpoint del backend que devuelva JSON tipado.</summary>
    public async Task<T?> PostAsync<T>(string path, object req, CancellationToken ct) where T : class
    {
        var body = JsonSerializer.Serialize(req, Json);
        using var content = new StringContent(body, Encoding.UTF8, "application/json");
        using var res = await _http.PostAsync($"{_baseUrl}{path}", content, ct);
        var text = await res.Content.ReadAsStringAsync(ct);
        if (!res.IsSuccessStatusCode)
            throw new InvalidOperationException($"backend HTTP {(int)res.StatusCode}: {text}");
        return JsonSerializer.Deserialize<T>(text, Json);
    }
}
